src/hotspot/share/opto/compile.cpp

```diff
@@ -4948,6 +4949,34 @@ bool Compile::coarsened_locks_consistent() {
   return true;
 }
 
+// Mark locking regions (identified by BoxLockNode) as coarsened
+// if locks coarsening optimization removed lock/unlock from them.
+// Such regions become unbalanced and we can't execute other
+// locks elimination optimization on them.
+void Compile::mark_coarsened_boxes() {
+  int count = coarsened_count();
+  for (int i = 0; i < count; i++) {
+    Node_List* locks_list = _coarsened_locks.at(i);
+    uint size = locks_list->size();
+    if (size > 0) {
+      AbstractLockNode* alock = locks_list->at(0)->as_AbstractLock();
+      BoxLockNode* box = alock->box_node()->as_BoxLock();
+      if (alock->is_coarsened() && !box->is_unbalanced()) { // Not marked already
+        assert(!box->is_eliminated(), "regions with coarsened locks should not be marked as eliminated");
+        for (uint j = 1; j < size; j++) {
+          assert(locks_list->at(j)->as_AbstractLock()->is_coarsened(), "only coarsened locks are expected here");
+          BoxLockNode* this_box = locks_list->at(j)->as_AbstractLock()->box_node()->as_BoxLock();
+          assert(!this_box->is_eliminated(), "regions with coarsened locks should not be marked as eliminated");
+          if (box != this_box) {
+            box->set_unbalanced();
+            this_box->set_unbalanced();
+          }
+        }
+      }
+    }
+  }
+}
+
```

Comment:
A code-style suggestion:
You use lots of nesting here, with deeper and deeper if blocks. But since you don't have any else-blocks, you could easily flatten your code by using `if (condition) { continue; }`, what I call "bailout instead of nesting". But that may be up to personal taste ;)

Reply:
I like nesting - it clear shows conditions under which nested code is executed. And it has less lines.
I don't think this code has deep nesting. If it was more than 10 nesting I would agree to your suggestion.